p2p: Correct unrealistic headerssync unit test behavior

hodlinator opened the pull request:
### Background

As part of the release process we often run *contrib/devtools/headerssync-params.py* and increase the values of the constants `HEADER_COMMITMENT_PERIOD` and `REDOWNLOAD_BUFFER_SIZE` in *src/headerssync.cpp* as per *doc/release-process.md* (example: 11a2d3a63e90cdc1920ede3c67d52a9c72860e6b). This helps fine tune the memory consumption per `HeadersSyncState`-instance in the face of malicious peers.

(The `REDOWNLOAD_BUFFER_SIZE`/`HEADER_COMMITMENT_PERIOD` ratio determines how many Headers Sync commitment bits must match between PRESYNC & REDOWNLOAD phases before we start permanently storing headers from a peer. For more details see comments in *src/headerssync.h* and *contrib/devtools/headerssync-params.py*).


### Problem: Not feeding back headers until completing sync

During v30 release process #33274 made `REDOWNLOAD_BUFFER_SIZE` exceed the `target_blocks` constant used to control the length of chains generated for testing Headers Sync (`15000`, *headers_sync_chainwork_tests.cpp*).

The `HeadersSyncState::m_redownloaded_headers`-buffer now does not reach the `REDOWNLOAD_BUFFER_SIZE`-threshold during those unit tests. As a consequence `HeadersSyncState::PopHeadersReadyForAcceptance()` will not start feeding back headers until the PoW threshold has been met. While this will not cause the unit test to start failing on master, it means we have gone from testing behavior that resembles mainnet (way more than `REDOWNLOAD_BUFFER_SIZE` headers to reach the PoW limit), to behavior that is not possible/expected there.


### Solution

Avoid testing this unrealistic condition of completing Headers Sync before reaching `REDOWNLOAD_BUFFER_SIZE` by making tests able to define their own values through the new `HeadersSyncParams` instead of having them hard-coded for all chains & tests.


### Commits

* First 6 commits refactor and improve the unit tests in order to clarify latter changes.
* We then add checks for the behavior around the `REDOWNLOAD_BUFFER_SIZE` threshold.
* The main change: we extract the section from *headerssync.cpp* containing the constants to *kernel/chainparams.cpp*, making `HeadersSyncState` no longer hard-coded to mainnet.


### Notes

This PR used to be called "headerssync: Preempt unrealistic unit test behavior".

l0rinc reviewed (changes requested):
Concept ACK – this PR replaces the imminently outdated hard-coded test target of 15'000 headers with a value derived from the live `REDOWNLOAD_BUFFER_SIZE` constant.
Moving both header-sync tuning constants into a shared header, and making the unit test reference them, keeps the test exercising the 'buffer-full' path as the parameters grow each release.

Recommendations:
* split the series into more commits, separating low-risk refactors from ones that change behaviour: strict move-only; refactor / cleanup with no behavioural change; extra assertions; wiring the test to the new parameter header;
* in the test file, wrap each independent `ProcessNextHeaders` call inside `{}` blocks and use structured bindings so temporary variables aren’t reused accidentally;
* rename the new file to `headerssync_params.h` (underscore instead of `-`) for consistency with other similar headers;
* update `contrib/devtools/README.md` to reference the header instead of the .cpp;
* Tweak a few comments whose wording is unclear or slightly misleading.

hodlinator reviewed (commented):
Thanks for your feedback so far @l0rinc! Took most of your suggestions (https://github.com/bitcoin/bitcoin/pull/32579#pullrequestreview-2857912770).

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -63,99 +65,107 @@ BOOST_FIXTURE_TEST_SUITE(headers_sync_chainwork_tests, HeadersGeneratorSetup)
 // 2. Then we deliver the second set of headers and verify that they fail
 //    processing (presumably due to commitments not matching).
-static void SneakyRedownload(const std::vector<CBlockHeader>& first_chain, const std::vector<CBlockHeader>& second_chain, const CBlockIndex* chain_start, const arith_uint256& chain_work);
+static void SneakyRedownload(const std::vector<CBlockHeader>& first_chain, const std::vector<CBlockHeader>& second_chain, const CBlockIndex* chain_start);
```
* they're simpler/less specific
* they're shorter (matters for these long lines)
* you're already changing them in other places

i.e. what's the point of keeping them as vectors when we don't actually need so much power?

l0rinc commented on src/arith_uint256.h:
Based on the [Sonarcloud](https://corecheck.dev/bitcoin/bitcoin/pulls/32579) recommendations I eagerly checked which other constructors can be explicit, but [exploring](https://cplusplus.com/forum/general/168292/) your question in more detail it seems there's barely any advantage in explicit default constructors.

jonatack commented:
Concept ACK

hodlinator commented on src/headerssync.cpp:
```diff
@@ -1,26 +1,18 @@
+// CompressedHeader (we should re-calculate parameters if we compress further).
 static_assert(sizeof(CompressedHeader) == 48);
 
 HeadersSyncState::HeadersSyncState(NodeId id, const Consensus::Params& consensus_params,
```
Just realized a different possible approach: passing `HEADER_COMMITMENT_PERIOD` and `REDOWNLOAD_BUFFER_SIZE` values into the `HeadersSyncState`-ctor. Right now that aspect of the class is hardcoded for mainnet, despite the ctor accepting `Consensus::Params`. Having these values be sent in would enable tests to send in different (smaller) variables for regtest, while still testing behavior around the threshold.

In fact, it would fit nicely with other values updated during the release process to have them written into the corresponding section (mainnet/regtest/etc) of *src/kernel/chainparams.cpp*.

Started implementing that approach to see how it pans out.

hodlinator reviewed (commented):
Latest push (42e746096b93b8026569817d8925750089d6daf4, [compare](https://github.com/bitcoin/bitcoin/compare/bbe5609b84211c49a522da9dd5d987ab454ef912..42e746096b93b8026569817d8925750089d6daf4)):
* Relaxes the hardcoding of the header commitment period and the redownload buffer size in `HeadersSyncState`.
  - This makes things more consistent with the constructor already taking `Consensus::Params` (implying different chains are supported).
  - Moves specification of the above constants from the top of *headerssync.cpp* into *kernel/chainparams.cpp* (similar to many other constants).
  - Returns the chainwork test suite back to only using an integer literal for `TARGET_BLOCKS` (instead of depending on an #included constant). Also now uses test-local values for commitment period and redownload buffer size.
* Adds a warning in `HeadersSyncState::PopHeadersReadyForAcceptance()` for the case when we reach minimum PoW before exceeding the redownload buffer size together with tests.

(Updated PR description).

l0rinc commented on src/headerssync.cpp:
```diff
@@ -278,6 +278,14 @@ std::vector<CBlockHeader> HeadersSyncState::PopHeadersReadyForAcceptance()
 
+    if (m_process_all_remaining_headers && m_chain_start->nHeight == 0 &&
+        m_redownload_buffer_first_prev_hash == m_chain_start->GetBlockHash() &&
+        m_redownloaded_headers.size() <= m_params.redownload_buffer_size) {
```
Q: I can't immediately tell if this should be triggered for equality as well - please just resolve if it's a stupid question.

l0rinc replied:
Hmmm, if it has to be the opposite of the while below, would it make sense to extract the common parts to explain it with code (your explanation makes sense, would prefer if the code already made it simpler so that I don't have to check the code that's after the current one to understand it).

hodlinator replied:
Broke out `bool buffer_exceeded`, let me know if you think it's an improvement.

hodlinator replied:
Latest push improves this part of the code, please check it out!

l0rinc reviewed (changes requested):
I might have gone a bit overboard with the test review, but I think we can make it a lot less noisy - here's the patch of many of my suggestions for convenience:

<details>
<summary>Patch</summary>

```patch
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index d64dd43aba..658ea7f2c3 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -193,8 +193,7 @@ public:
         // Generated by headerssync-params.py on 2025-03-04.
         m_headers_sync_params = HeadersSyncParams{
             .commitment_period = 624,
-            .redownload_buffer_size = 14827, // 14827/624 = ~23.8 commitments
-        };
+            .redownload_buffer_size = 14827}; // ~24 commitments
     }
 };
 
diff --git a/src/test/headers_sync_chainwork_tests.cpp b/src/test/headers_sync_chainwork_tests.cpp
index 1f85c69838..ccbbc1be50 100644
--- a/src/test/headers_sync_chainwork_tests.cpp
+++ b/src/test/headers_sync_chainwork_tests.cpp
@@ -13,10 +13,13 @@
 #include <validation.h>
 
 #include <cstddef>
+#include <optional>
 #include <vector>
 
 #include <boost/test/unit_test.hpp>
 
+using State = HeadersSyncState::State;
+
 constexpr size_t TARGET_BLOCKS{15'000};
 constexpr arith_uint256 CHAIN_WORK{TARGET_BLOCKS * 2};
 constexpr size_t REDOWNLOAD_BUFFER_SIZE{TARGET_BLOCKS - (MAX_HEADERS_RESULTS + 123)};
@@ -37,18 +40,18 @@ static void FindProofOfWork(CBlockHeader& starting_header)
  * prev_time, and with a fixed merkle root hash.
  */
 static void GenerateHeaders(std::vector<CBlockHeader>& headers,
-        size_t count, const uint256& starting_hash, const int nVersion, int prev_time,
-        const uint256& merkle_root, const uint32_t nBits)
+                            size_t count, const uint256& starting_hash, int nVersion, int prev_time,
+                            const uint256& merkle_root, uint32_t nBits)
 {
-    uint256 prev_hash = starting_hash;
+    uint256 prev_hash{starting_hash};
 
     while (headers.size() < count) {
         headers.emplace_back();
-        CBlockHeader& next_header = headers.back();
+        CBlockHeader& next_header{headers.back()};
         next_header.nVersion = nVersion;
         next_header.hashPrevBlock = prev_hash;
         next_header.hashMerkleRoot = merkle_root;
-        next_header.nTime = prev_time+1;
+        next_header.nTime = prev_time + 1;
         next_header.nBits = nBits;
 
         FindProofOfWork(next_header);
@@ -57,220 +60,167 @@ static void GenerateHeaders(std::vector<CBlockHeader>& headers,
     }
 }
 
-static HeadersSyncState CreateState(const CBlockIndex* chain_start, const CChainParams& chain_params)
+static HeadersSyncState CreateState(const CChainParams& chain_params, const HeadersSyncParams& sync_params, const CBlockIndex* chain_start)
 {
     return {/*id=*/0,
             chain_params.GetConsensus(),
-            HeadersSyncParams{
-                .commitment_period = COMMITMENT_PERIOD,
-                .redownload_buffer_size = REDOWNLOAD_BUFFER_SIZE,
-            },
+            sync_params,
             chain_start,
-            /*minimum_required_work=*/CHAIN_WORK};
+           /*minimum_required_work=*/CHAIN_WORK};
 }
 
-BOOST_FIXTURE_TEST_SUITE(headers_sync_chainwork_tests, RegTestingSetup)
-
-// In this test, we construct two sets of headers from genesis, one with
-// sufficient proof of work and one without.
-// 1. We deliver the first set of headers and verify that the headers sync state
-//    updates to the REDOWNLOAD phase successfully.
-// 2. Then we deliver the second set of headers and verify that they fail
-//    processing (presumably due to commitments not matching).
-static void SneakyRedownload(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain,
-        const std::vector<CBlockHeader>& second_chain);
-// 3. Verify that repeating with the first set of headers in both phases is
-//    successful.
-static void HappyPath(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain);
-// 4. Finally, repeat the second set of headers in both phases to demonstrate
-//    behavior when the chain a peer provides has too little work.
-static void TooLittleWork(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& second_chain);
-
-static void TooBigBuffer(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain);
-
-BOOST_AUTO_TEST_CASE(headers_sync_state)
+// Standard set of checks common to all scenarios; macro keeps failure lines at the call-site.
+#define CHECK_RESULT(res, state, exp_state, exp_success, exp_request_more, exp_headers_size) \
+    do {                                                                                     \
+        const auto& [res_headers, res_success, res_request_more]{res};                       \
+        BOOST_REQUIRE_EQUAL(state.GetState(), exp_state);                                    \
+        BOOST_CHECK_EQUAL(res_headers.size(), exp_headers_size);                             \
+        BOOST_CHECK_EQUAL(res_success, exp_success);                                         \
+        BOOST_CHECK_EQUAL(res_request_more, exp_request_more);                               \
+    } while (false)
+
+// Deliver the first set of headers from genesis to reach REDOWNLOAD,
+// then deliver the second set and verify that commitments don’t match.
+static void SneakyRedownload(const CBlockIndex* chain_start, const std::vector<CBlockHeader>& chain1, const std::vector<CBlockHeader>& chain2)
 {
-    std::vector<CBlockHeader> first_chain;
-    std::vector<CBlockHeader> second_chain;
-
-    const auto genesis{Params().GenesisBlock()};
-
-    // Generate headers for two different chains (using differing merkle roots
-    // to ensure the headers are different).
-    GenerateHeaders(first_chain, TARGET_BLOCKS - 1, genesis.GetHash(), genesis.nVersion,
-                    genesis.nTime, /*merkle_root=*/uint256::ZERO, genesis.nBits);
-    GenerateHeaders(second_chain, TARGET_BLOCKS - 2, genesis.GetHash(), genesis.nVersion,
-                    genesis.nTime, /*merkle_root=*/uint256::ONE, genesis.nBits);
-
-    const CBlockIndex* chain_start = WITH_LOCK(::cs_main, return m_node.chainman->m_blockman.LookupBlockIndex(genesis.GetHash()));
+    HeadersSyncState state{CreateState(Params(), HeadersSyncParams{COMMITMENT_PERIOD, REDOWNLOAD_BUFFER_SIZE}, chain_start)};
+    const auto genesis_hash{Params().GenesisBlock().GetHash()};
 
-    SneakyRedownload(chain_start, first_chain, second_chain);
-    HappyPath(chain_start, first_chain);
-    TooLittleWork(chain_start, second_chain);
-    TooBigBuffer(chain_start, first_chain);
+    const std::span headers{chain1};
+    CHECK_RESULT(state.ProcessNextHeaders(headers.first(1), /*full_headers_message=*/true),
+                 state, State::PRESYNC,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+    BOOST_CHECK_EQUAL(state.NextHeadersRequestLocator().vHave.front(), chain1.front().GetHash());
+
+    // Pretend the message is still "full", so we don't abort.
+    // This chain should look valid, and we should have met the proof-of-work
+    // requirement during PRESYNC and transitioned to REDOWNLOAD.
+    CHECK_RESULT(state.ProcessNextHeaders(headers.last(headers.size() - 1), /*full_headers_message=*/true),
+                 state, State::REDOWNLOAD,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+    BOOST_CHECK_EQUAL(state.NextHeadersRequestLocator().vHave.front(), genesis_hash);
+
+    // Try to sneakily feed back the second chain during REDOWNLOAD.
+    CHECK_RESULT(state.ProcessNextHeaders(chain2, /*full_headers_message=*/true),
+                 state, State::FINAL,
+                 /*exp_success=*/false,
+                 /*exp_request_more=*/false,
+                 /*exp_headers_size=*/0);
 }
 
-static void SneakyRedownload(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain,
-        const std::vector<CBlockHeader>& second_chain)
+// Verify that repeating with the first set of headers in both phases is successful.
+static void HappyPath(const CBlockIndex* chain_start, const std::vector<CBlockHeader>& chain1)
 {
-    // Feed the first chain to HeadersSyncState, by delivering 1 header
-    // initially and then the rest.
-    HeadersSyncState hss{CreateState(chain_start, Params())};
-    {
-        // Just feed one header and check state.
-        auto result{hss.ProcessNextHeaders(std::span{first_chain.begin(), 1}, true)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::PRESYNC);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-        BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.front(), first_chain.front().GetHash());
-    }
-
-    {
-        // Pretend the message is still "full", so we don't abort.
-        auto result{hss.ProcessNextHeaders(std::span{first_chain.begin() + 1, first_chain.end()}, true)};
-        // This chain should look valid, and we should have met the proof-of-work
-        // requirement during PRESYNC and transitioned to REDOWNLOAD.
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::REDOWNLOAD);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-        // The locator should reset to genesis.
-        BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.front(), Params().GenesisBlock().GetHash());
-    }
+    // Feed the first chain twice.
+    HeadersSyncState state{CreateState(Params(), HeadersSyncParams{COMMITMENT_PERIOD, REDOWNLOAD_BUFFER_SIZE}, chain_start)};
+    const auto genesis_hash{Params().GenesisBlock().GetHash()};
+    const std::span headers{chain1};
 
-    {
-        // Try to sneakily feed back the second chain during REDOWNLOAD.
-        auto result{hss.ProcessNextHeaders(second_chain, true)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::FINAL);
-        BOOST_CHECK(!result.success); // Foiled! We detected mismatching headers.
-        BOOST_CHECK(!result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-    }
+    // During normal operation we shouldn't get the redownload buffer size warning.
+    ASSERT_NO_DEBUG_LOG(BUFFER_SIZE_WARNING);
+
+    // Switched from PRESYNC to REDOWNLOAD after reaching sufficient work:
+    CHECK_RESULT(state.ProcessNextHeaders(headers, /*full_headers_message=*/true),
+                 state, State::REDOWNLOAD,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+    BOOST_CHECK_EQUAL(state.NextHeadersRequestLocator().vHave.front(), genesis_hash);
+
+    // Process only so that the internal threshold isn't met, meaning validated headers shouldn't be returned yet:
+    CHECK_RESULT(state.ProcessNextHeaders(headers.first(REDOWNLOAD_BUFFER_SIZE), /*full_headers_message=*/true),
+                 state, State::REDOWNLOAD,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+    BOOST_CHECK_EQUAL(state.NextHeadersRequestLocator().vHave.front(), chain1[REDOWNLOAD_BUFFER_SIZE - 1].GetHash());
+
+    // Next header should make us exceed the threshold, but still not be done:
+    const auto res_threshold{state.ProcessNextHeaders(headers.subspan(REDOWNLOAD_BUFFER_SIZE, 1), /*full_headers_message=*/true)};
+    CHECK_RESULT(res_threshold,
+                 state, State::REDOWNLOAD,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/1);
+    BOOST_CHECK_EQUAL(state.NextHeadersRequestLocator().vHave.front(), chain1[REDOWNLOAD_BUFFER_SIZE].GetHash());
+
+    // Feed in remaining headers, meeting the work threshold again and completing the REDOWNLOAD phase:
+    CHECK_RESULT(state.ProcessNextHeaders(headers.subspan(REDOWNLOAD_BUFFER_SIZE + 1), /*full_headers_message=*/true),
+                 state, State::FINAL,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/false,
+                 /*exp_headers_size=*/chain1.size() - 1);
+    BOOST_CHECK_EQUAL(res_threshold.pow_validated_headers.front().hashPrevBlock, genesis_hash);
 }
 
-static void HappyPath(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain)
+// Repeat the second set of headers in both phases to demonstrate
+// behavior when the chain a peer provides has too little work.
+static void TooLittleWork(const CBlockIndex* chain_start, const std::vector<CBlockHeader>& chain2)
 {
-    // This time we feed the first chain twice.
-    HeadersSyncState hss{CreateState(chain_start, Params())};
-
-    // During normal operation we shouldn't get the redownload buffer size warning.
-    DebugLogHelper forbidden{BUFFER_SIZE_WARNING, [](const std::string* str) {
-        if (str == nullptr) {
-            return false; // Disable exception for not finding a match.
-        } else {
-            throw std::runtime_error{strprintf("'%s' should not exist in debug log\n", str)};
-        }
-    }};
-
-    const auto genesis_hash{Params().GenesisBlock().GetHash()};
-    {
-        auto result{hss.ProcessNextHeaders(first_chain, true)};
-        // Switched from PRESYNC to REDOWNLOAD after reaching sufficient work:
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::REDOWNLOAD);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-        // The locator should reset to genesis.
-        BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.front(), genesis_hash);
-    }
-
-    {
-        // Process only so that the internal threshold isn't met, meaning validated
-        // headers shouldn't be returned yet:
-        auto result{hss.ProcessNextHeaders(std::span{first_chain.begin(), REDOWNLOAD_BUFFER_SIZE}, true)};
-        // Not done:
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::REDOWNLOAD);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-        BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.front(), first_chain[REDOWNLOAD_BUFFER_SIZE - 1].GetHash());
-    }
-
-    CBlockHeader first_after_genesis;
-    {
-        // Next header should make us exceed the threshold, but still not be done:
-        auto result{hss.ProcessNextHeaders(std::span{first_chain.begin() + REDOWNLOAD_BUFFER_SIZE, 1}, true)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::REDOWNLOAD);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_REQUIRE_EQUAL(result.pow_validated_headers.size(), 1);
-        BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.front(), first_chain[REDOWNLOAD_BUFFER_SIZE].GetHash());
-        first_after_genesis = result.pow_validated_headers.front();
-        BOOST_CHECK_EQUAL(first_after_genesis.hashPrevBlock, genesis_hash);
-    }
-
-    {
-        // Feed in remaining headers, meeting the work threshold again and
-        // completing the REDOWNLOAD phase.
-        auto result{hss.ProcessNextHeaders(std::span{first_chain.begin() + REDOWNLOAD_BUFFER_SIZE + 1, first_chain.end()}, true)};
-        // Nothing left for the sync logic to do:
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::FINAL);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(!result.request_more);
-        // All headers except the one already returned above:
-        BOOST_REQUIRE_EQUAL(result.pow_validated_headers.size(), first_chain.size() - 1);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.front().hashPrevBlock, first_after_genesis.GetHash());
-    }
+    // Verify that just trying to process the second chain would not succeed (too little work).
+    HeadersSyncState state{CreateState(Params(), HeadersSyncParams{COMMITMENT_PERIOD, REDOWNLOAD_BUFFER_SIZE}, chain_start)};
+    const std::span headers{chain2};
+
+    // Pretend just the first message is "full", so we don't abort.
+    CHECK_RESULT(state.ProcessNextHeaders(headers.first(1), /*full_headers_message=*/true),
+                 state, State::PRESYNC,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+
+    // Non-full message causes sync to end with no headers accepted.
+    CHECK_RESULT(state.ProcessNextHeaders(headers.subspan(1), /*full_headers_message=*/false),
+                 state, State::FINAL,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/false,
+                 /*exp_headers_size=*/0);
 }
 
-static void TooLittleWork(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& second_chain)
+// Scenario that intentionally uses an oversized buffer to trigger the warning.
+static void TooBigBuffer(const CBlockIndex* chain_start, const std::vector<CBlockHeader>& chain1)
 {
-    // Verify that just trying to process the second chain would not succeed
-    // (too little work).
-    HeadersSyncState hss{CreateState(chain_start, Params())};
-    BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::PRESYNC);
-    {
-        // Pretend just the first message is "full", so we don't abort.
-        auto result{hss.ProcessNextHeaders(std::span{second_chain.begin(), 1}, true)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::PRESYNC);
-        BOOST_CHECK(result.success);
-        BOOST_CHECK(result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-    }
-
-    {
-        // Tell the sync logic that the headers message was not full, implying no
-        // more headers can be requested. For a low-work-chain, this should cause
-        // the sync to end with no headers for acceptance.
-        auto result{hss.ProcessNextHeaders(std::span{second_chain.begin() + 1, second_chain.end()}, false)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::FINAL);
-        BOOST_CHECK(!result.request_more);
-        BOOST_CHECK_EQUAL(result.pow_validated_headers.size(), 0);
-        // Nevertheless, no validation errors should have been detected with the
-        // chain:
-        BOOST_CHECK(result.success);
-    }
+    // Intentionally too big redownload buffer in order to trigger warning.
+    HeadersSyncState state{CreateState(Params(), HeadersSyncParams{COMMITMENT_PERIOD, chain1.size()}, chain_start)};
+
+    CHECK_RESULT(state.ProcessNextHeaders(chain1, /*full_headers_message=*/true),
+                 state, State::REDOWNLOAD,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/true,
+                 /*exp_headers_size=*/0);
+
+    ASSERT_DEBUG_LOG(BUFFER_SIZE_WARNING);
+    CHECK_RESULT(state.ProcessNextHeaders(chain1, /*full_headers_message=*/true),
+                 state, State::FINAL,
+                 /*exp_success=*/true,
+                 /*exp_request_more=*/false,
+                 /*exp_headers_size=*/chain1.size());
 }
 
-static void TooBigBuffer(const CBlockIndex* chain_start,
-        const std::vector<CBlockHeader>& first_chain)
+BOOST_FIXTURE_TEST_SUITE(headers_sync_chainwork_tests, RegTestingSetup)
+
+BOOST_AUTO_TEST_CASE(headers_sync_state)
 {
-    // Intentionally too big redownload buffer in order to trigger warning.
-    HeadersSyncState hss{/*id=*/0,
-                         Params().GetConsensus(),
-                         HeadersSyncParams{
-                             .commitment_period = COMMITMENT_PERIOD,
-                             .redownload_buffer_size = first_chain.size(),
-                         },
-                         chain_start,
-                         /*minimum_required_work=*/CHAIN_WORK};
-    (void)hss.ProcessNextHeaders(first_chain, true);
-    BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::REDOWNLOAD);
-
-    {
-        ASSERT_DEBUG_LOG(BUFFER_SIZE_WARNING);
-        auto result{hss.ProcessNextHeaders(first_chain, true)};
-        BOOST_REQUIRE_EQUAL(hss.GetState(), HeadersSyncState::State::FINAL);
-        BOOST_CHECK(result.success);
-    }
+    const auto& genesis{Params().GenesisBlock()};
+    const uint256 genesis_hash{genesis.GetHash()};
+
+    // Generate headers for two different chains (using differing merkle roots
+    // to ensure the headers are different).
+    std::vector<CBlockHeader> chain1{};
+    GenerateHeaders(chain1, TARGET_BLOCKS - 1, genesis_hash, genesis.nVersion, genesis.nTime,
+                    /*merkle_root=*/uint256::ZERO, genesis.nBits);
+    std::vector<CBlockHeader> chain2{};
+    GenerateHeaders(chain2, TARGET_BLOCKS - 2, genesis_hash, genesis.nVersion, genesis.nTime,
+                    /*merkle_root=*/uint256::ONE, genesis.nBits);
+
+    const CBlockIndex* chain_start{WITH_LOCK(::cs_main, return m_node.chainman->m_blockman.LookupBlockIndex(genesis_hash))};
+    SneakyRedownload(chain_start, chain1, chain2);
+    HappyPath(chain_start, chain1);
+    TooLittleWork(chain_start, chain2);
+    TooBigBuffer(chain_start, chain1);
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/util/logging.h b/src/test/util/logging.h
index 73ac23825f..622ff58938 100644
--- a/src/test/util/logging.h
+++ b/src/test/util/logging.h
@@ -36,6 +36,16 @@ public:
     ~DebugLogHelper() { check_found(); }
 };
 
+// Test fails if the pattern *DOES NOT* show up.
 #define ASSERT_DEBUG_LOG(message) DebugLogHelper UNIQUE_NAME(debugloghelper)(message)
+// Test fails if the pattern *DOES* show up.
+#define ASSERT_NO_DEBUG_LOG(message)                                              \
+    DebugLogHelper UNIQUE_NAME(nologhelper){                                      \
+        message,                                                                  \
+        [](const std::string* line) {                                             \
+            if (line) throw std::runtime_error("unexpected log line: " + *line);  \
+            return false; /* suppress default 'not found' failure */              \
+        }                                                                         \
+    }
 
 #endif // BITCOIN_TEST_UTIL_LOGGING_H
```


</details>

hodlinator reviewed (commented):
Thanks for the extensive review @l0rinc! Implemented many of the suggestions and answered the other points.

Removed the commit breaking out scopes in favor of the suggested `CHECK_RESULT` macro (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2158508969). Broke out the comment improvements into their own commit ahead of that since it gives a "skeleton" for the next commit's diff to hang off of.

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -9,39 +9,37 @@
+constexpr size_t TARGET_BLOCKS{15'000};
+constexpr arith_uint256 CHAIN_WORK{TARGET_BLOCKS * 2};
+
+/** Search for a nonce to meet (regtest) proof of work */
```
ah, so it's jut a reassurance that it doesn't need a fancy asic, it can mine a block in no time - makes sense

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -15,9 +15,36 @@
+        if (locator_hash_opt == std::nullopt) {                                                            \
+            BOOST_CHECK_EQUAL(exp_state, State::FINAL);                                                    \
+        } else {                                                                                           \
+            BOOST_REQUIRE_GT(hss.NextHeadersRequestLocator().vHave.size(), 0);                             \
```
would't an `.at(0)` suffice?

l0rinc replied:
I think it applies to `BOOST_REQUIRE_GT(result.pow_validated_headers.size(), 0);` as well

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -130,16 +165,20 @@ static void HappyPath(const CBlockIndex* chain_start,
+    CHECK_RESULT(hss.ProcessNextHeaders(first_chain, /*full_headers_message=*/true),
+        hss, /*exp_state=*/State::REDOWNLOAD,
+        /*exp_success*/true, /*exp_request_more=*/true,
+        // The locator should reset to genesis.
```
this comment is obvious from the code now: `/*exp_locator_hash=*/Params().GenesisBlock().GetHash()` or `/*exp_locator_hash=*/genesis_hash` later in 65a96b507c11b3f34efa996919892fa1a2fcf49c

l0rinc commented on contrib/devtools/headerssync-params.py:
```diff
@@ -337,15 +337,15 @@ def analyze(when):
+    print( "        m_headers_sync_params = HeadersSyncParams{")
+    print(f"            .commitment_period = {period},")
+    print(f"            .redownload_buffer_size = {bufsize},"
           f" // {bufsize}/{period} = ~{bufsize/period:.1f} commitments")
```
> makes it more likely that the number changes between releases

That sounds like a counter-argument to me, but don't have strong feelings about it, if you disagree, I don't mind, please resolve it.

l0rinc commented on src/kernel/chainparams.h:
```diff
@@ -61,6 +61,17 @@ struct ChainTxData {
+struct HeadersSyncParams {
+    //! Distance in blocks between header commitments.
+    size_t commitment_period;
+    //! Only feed headers to validation once this many headers on top have been
+    //! received and validated against commitments.
```
I'm not sure what "this many headers *on top* have been received"

hodlinator replied:
Agree. New attempt:
```C++
//! Only start outputting headers once this many headers have been received
//! and validated against commitments.
```

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -172,19 +172,37 @@ static void HappyPath(const CBlockIndex* chain_start,
+        // Not done:
+        hss, /*exp_state=*/State::REDOWNLOAD,
+        /*exp_success*/true, /*exp_request_more=*/true,
+        // The locator should reset to genesis.
```
is this comment accurate?

hodlinator replied:
No, rebase mishap. Thanks for catching!

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -109,6 +112,9 @@ static void HappyPath(const CBlockIndex*, const std::vector<CBlockHeader>&);
 // 4. Repeat the second set of headers in both phases to demonstrate behavior
 //    when the chain a peer provides has too little work.
 static void TooLittleWork(const CBlockIndex*, const std::vector<CBlockHeader>&);
+// 5. Sets the redownload buffer size to be large enough large enough that we
```
```suggestion
// 5. Sets the redownload buffer size to be large enough that we
```

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -234,4 +244,28 @@ static void TooLittleWork(const CBlockIndex* chain_start,
+static void TooBigBuffer(const CBlockIndex* chain_start,
+        const std::vector<CBlockHeader>& first_chain)
+{
+    // Intentionally too big redownload buffer in order to trigger warning.
```
nice, this is exactly how comments should be used to augment what the code cannot easily tell <3

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -234,4 +244,28 @@ static void TooLittleWork(const CBlockIndex* chain_start,
+        hss, /*exp_state=*/State::REDOWNLOAD,
+        /*exp_success*/true, /*exp_request_more=*/true,
+        /*exp_headers_size=*/0, /*exp_pow_validated_prev=*/std::nullopt,
+        /*exp_locator_hash=*/Params().GenesisBlock().GetHash());
```
might make sense to extract genesis hash here as well like we did in `HappyPath`

l0rinc reviewed (approved):
I find the current layout of tiny, focused commits a lot easier to follow, thanks!
Left a few comments, after that I will test it again locally and I can ACK.

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -60,15 +60,23 @@ BOOST_FIXTURE_TEST_SUITE(headers_sync_chainwork_tests, RegTestingSetup)
 //    processing (presumably due to commitments not matching).
-// 3. Finally, we verify that repeating with the first set of headers in both
-//    phases is successful.
+static void SneakyRedownload(const CBlockIndex* chain_start,
```
Can we add 2 helpers in that case? I really dislike that the comments don't follow the code's structure :.

hodlinator replied:
Renumbered in the "Improve comments"-commit.

danielabrozzoni commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -15,9 +15,34 @@
+        } else {                                                                                         \
+            BOOST_CHECK_EQUAL(hss.NextHeadersRequestLocator().vHave.at(0), locator_hash_opt);            \
+        }                                                                                                \
+        std::optional<uint256> pow_validated_prev_opt{exp_pow_validated_prev};                           \
```
micro nit in 79bdc6a785bca96dababc698336c04b8625b7d1c: you could check `exp_pow_validated_prev` first and `exp_locator_hash` second, to maintain the same order as the macro parameters

hodlinator replied:
Agree that's cleaner, fixed!
Also changed from
`== std::nullopt` to
`.has_value()`.

danielabrozzoni reviewed (commented):
Concept ACK, I've left a few questions/nits

I've only done a partial review so far, I still need to go through the last two commits and part of ced2ddb37c957e53d4a8aa40d3512ea3786511b4

l0rinc commented on src/arith_uint256.h:
```diff
@@ -239,11 +239,12 @@ class base_uint
-    arith_uint256(const base_uint<256>& b) : base_uint<256>(b) {}
-    arith_uint256(uint64_t b) : base_uint<256>(b) {}
+    constexpr arith_uint256() = default;
+    constexpr arith_uint256(const base_uint& b) : base_uint(b) {}
+    constexpr arith_uint256(uint64_t b) : base_uint(b) {}
```
Yes, they're redundant, can be deduced by the compiler

hodlinator commented on src/kernel/chainparams.h:
```diff
@@ -61,6 +61,17 @@ struct ChainTxData {
+struct HeadersSyncParams {
+    //! Distance in blocks between header commitments.
+    size_t commitment_period;
+    //! Only start outputting headers once this many headers have been received
```
Changed to the following, let me know what you have further suggestions:
> Start feeding back headers into the permanent block index once more than this number of them have been received and validated against commitments.

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -105,22 +130,29 @@ static void SneakyRedownload(const CBlockIndex* chain_start,
-    BOOST_CHECK(result.success);
-    BOOST_CHECK(result.request_more);
-    BOOST_CHECK(hss.GetState() == HeadersSyncState::State::REDOWNLOAD);
+    CHECK_RESULT(hss.ProcessNextHeaders(std::span{first_chain}.last(first_chain.size() - 1), true),
```
In this case we have the comment right above describing the same thing:
https://github.com/bitcoin/bitcoin/blob/79bdc6a785bca96dababc698336c04b8625b7d1c/src/test/headers_sync_chainwork_tests.cpp#L140

Went through the code and realized `HappyPath` was repeating `/*full_headers_message=*/` in one piece of newer code, so removed that.

This boolean parameter to `ProcessNextHeaders()` is somewhat of a code smell. Maybe could be split into 2 public functions (`ProcessNextHeaders()` + `ProcessLastHeaders()`... but that would make the code as a whole more convoluted I think).

hodlinator reviewed (commented):
Thanks for checking this out @danielabrozzoni! Pushed some minor changes based on your feedback.

danielabrozzoni reviewed (commented):
Review 78503a41546f6432045d3f9fd39860ead11d5c95

Code looks good to me, tests are passing locally. Thanks for picking up my comments! I left a couple more.

I did some additional testing by cherry-picking the latest commit of this PR and applying it to master, setting REDOWNLOAD_BUFFER_SIZE to >15000, and verifying that the warning was logged as expected.

<details>
<summary> git diff: </summary>

```diff
diff --git a/src/headerssync.cpp b/src/headerssync.cpp
index 9e8b190516..0d0a26d5db 100644
--- a/src/headerssync.cpp
+++ b/src/headerssync.cpp
@@ -17,7 +17,7 @@ constexpr size_t HEADER_COMMITMENT_PERIOD{624};
 
 //! Only feed headers to validation once this many headers on top have been
 //! received and validated against commitments.
-constexpr size_t REDOWNLOAD_BUFFER_SIZE{14827}; // 14827/624 = ~23.8 commitments
+constexpr size_t REDOWNLOAD_BUFFER_SIZE{15002}; // 14827/624 = ~23.8 commitments
 
 // Our memory analysis assumes 48 bytes for a CompressedHeader (so we should
 // re-calculate parameters if we compress further)
@@ -285,11 +285,20 @@ std::vector<CBlockHeader> HeadersSyncState::PopHeadersReadyForAcceptance()
     Assume(m_download_state == State::REDOWNLOAD);
     if (m_download_state != State::REDOWNLOAD) return ret;
 
-    while (m_redownloaded_headers.size() > REDOWNLOAD_BUFFER_SIZE ||
+    bool buffer_exceeded{m_redownloaded_headers.size() > REDOWNLOAD_BUFFER_SIZE};
+    if (!buffer_exceeded && m_process_all_remaining_headers && m_chain_start->nHeight == 0 &&
+            m_redownload_buffer_first_prev_hash == m_chain_start->GetBlockHash()) {
+        LogWarning("Unexpectedly reached minimum required work before filling the headers sync redownload buffer, even though we started all the way back from genesis "
+                   "(m_redownload_buffer_last_height: %u, redownload_buffer_size: %u).",
+                   m_redownload_buffer_last_height, REDOWNLOAD_BUFFER_SIZE);
+    }
+
+    while (buffer_exceeded ||
             (m_redownloaded_headers.size() > 0 && m_process_all_remaining_headers)) {
         ret.emplace_back(m_redownloaded_headers.front().GetFullHeader(m_redownload_buffer_first_prev_hash));
         m_redownloaded_headers.pop_front();
         m_redownload_buffer_first_prev_hash = ret.back().GetHash();
+        buffer_exceeded = m_redownloaded_headers.size() > REDOWNLOAD_BUFFER_SIZE;
     }
     return ret;
 }
diff --git a/src/test/headers_sync_chainwork_tests.cpp b/src/test/headers_sync_chainwork_tests.cpp
index b710ad1801..a14aa0ee6d 100644
--- a/src/test/headers_sync_chainwork_tests.cpp
+++ b/src/test/headers_sync_chainwork_tests.cpp
@@ -7,6 +7,7 @@
 #include <consensus/params.h>
 #include <headerssync.h>
 #include <pow.h>
+#include <test/util/logging.h>
 #include <test/util/setup_common.h>
 #include <validation.h>
 #include <vector>
@@ -65,8 +66,12 @@ BOOST_FIXTURE_TEST_SUITE(headers_sync_chainwork_tests, HeadersGeneratorSetup)
 //    processing (presumably due to commitments not matching).
 // 3. Finally, we verify that repeating with the first set of headers in both
 //    phases is successful.
+
+constexpr char BUFFER_SIZE_WARNING[]{"Unexpectedly reached minimum required work before filling the headers sync redownload buffer, even though we started all the way back from genesis"};
+
 BOOST_AUTO_TEST_CASE(headers_sync_state)
 {
+    ASSERT_NO_DEBUG_LOG(BUFFER_SIZE_WARNING);
     std::vector<CBlockHeader> first_chain;
     std::vector<CBlockHeader> second_chain;
 
diff --git a/src/test/util/logging.h b/src/test/util/logging.h
index 73ac23825f..af7110b279 100644
--- a/src/test/util/logging.h
+++ b/src/test/util/logging.h
@@ -36,6 +36,17 @@ public:
     ~DebugLogHelper() { check_found(); }
 };
 
+// Test fails if the pattern *DOES NOT* show up.
 #define ASSERT_DEBUG_LOG(message) DebugLogHelper UNIQUE_NAME(debugloghelper)(message)
 
+// Test fails if the pattern *DOES* show up.
+#define ASSERT_NO_DEBUG_LOG(message)                                                        \
+    DebugLogHelper UNIQUE_NAME(nologhelper){                                                \
+        message,                                                                            \
+        [](const std::string* line) {                                                       \
+            if (line) throw std::runtime_error{"Encountered forbidden log line: " + *line}; \
+            return false; /* Suppress default 'not found' failure. */                       \
+        }                                                                                   \
+    }
+
 #endif // BITCOIN_TEST_UTIL_LOGGING_H	
```
</details>

Test output:
```
./build/bin/test_bitcoin --run_test=headers_sync_chainwork_tests
Running 1 test case...
unknown location(0): fatal error: in "headers_sync_chainwork_tests/headers_sync_state": std::runtime_error: Encountered forbidden log line: 2025-07-09T14:20:58.691984Z [test] [headerssync.cpp:291] [PopHeadersReadyForAcceptance] [warning] Unexpectedly reached minimum required work before filling the headers sync redownload buffer, even though we started all the way back from genesis (m_redownload_buffer_last_height: 14999, redownload_buffer_size: 15002).

./test/headers_sync_chainwork_tests.cpp(119): last checkpoint

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
error: Recipe `unit-test` failed on line 13 with exit code 201

```

hodlinator reviewed (commented):
Thanks @danielabrozzoni for your latest feedback and testing against master with your diff! Added some comments in latest push.

danielabrozzoni commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -52,95 +80,188 @@ void HeadersGeneratorSetup::GenerateHeaders(std::vector<CBlockHeader>& headers,
-    GenerateHeaders(first_chain, target_blocks-1, Params().GenesisBlock().GetHash(),
-            Params().GenesisBlock().nVersion, Params().GenesisBlock().nTime,
-            ArithToUint256(0), Params().GenesisBlock().nBits);
+    GenerateHeaders(first_chain, TARGET_BLOCKS - 1, genesis.GetHash(), genesis.nVersion,
```
Thanks a lot for the explanation, and the updated comments! It's much more understandable now.

hodlinator reviewed (commented):
Changes with latest push ([compare](https://github.com/bitcoin/bitcoin/compare/37405238a8c4fe4dff82781c91fb668391eb51d4..538b9a4f4ee84749bd1b2fe32e61658c272ee7f9)):
* Clarified PR description, hopefully better motivation reason for v30 milestone inclusion.
* Moved checks for behavior around `pow_validated_headers` and `REDOWNLOAD_BUFFER_SIZE` to first commit to motivate the rest of the PR instead of being one of the last commits.
* Simplified condition emitting warning for misconfigured `REDOWNLOAD_BUFFER_SIZE`.

l0rinc reviewed (changes requested):
Code reviewed until 058ebc1b02ccfc75bbe5178a4191aebdcf9b4d3a

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -9,39 +9,37 @@
+constexpr arith_uint256 CHAIN_WORK{TARGET_BLOCKS * 2};
+
+/** Search for a nonce to meet (regtest) proof of work */
+static void FindProofOfWork(CBlockHeader& starting_header)
```
Name could be better, but keeping it for now as it helps make the diff somewhat less jarring.

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -124,9 +145,15 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
+{
+    std::unique_ptr<HeadersSyncState> hss;
+    HeadersSyncState::ProcessingResult result;
+    // Verify that just trying to process the second chain would not succeed
```
Postponed updating this and one other comment until the "Improve comments" commit.

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -72,27 +69,23 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
 
-    const int target_blocks = 15000;
-    arith_uint256 chain_work = target_blocks*2;
+    const auto genesis{Params().GenesisBlock()};
```
Added ampersand but keeping the 3 `GetHash()` calls for now.

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -113,11 +113,28 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
     BOOST_CHECK(!result.request_more);
     // All headers should be ready for acceptance:
-    BOOST_CHECK(result.pow_validated_headers.size() == first_chain.size());
+    BOOST_CHECK_EQUAL(validated_headers_count + result.pow_validated_headers.size(), first_chain.size());
```
(This is the check that fails for me specifically):
https://github.com/bitcoin/bitcoin/blob/57a351912b057d1a2234d4a3b3dd0c745690e3d0/src/test/headers_sync_chainwork_tests.cpp#L127
```
../src/test/headers_sync_chainwork_tests.cpp(127): error: in "headers_sync_chainwork_tests/headers_sync_state": check result.pow_validated_headers.size() > 0 has failed [0 <= 0]
```

marcofleon replied:
Same check fails for me too 🤝

hodlinator reviewed (commented):
Pushed updates based off https://github.com/bitcoin/bitcoin/pull/32579#pullrequestreview-3112310893.

l0rinc commented on src/headerssync.cpp:
```diff
@@ -193,7 +186,7 @@ bool HeadersSyncState::ValidateAndProcessSingleHeader(const CBlockHeader& curren
     }
 
-    if (next_height % HEADER_COMMITMENT_PERIOD == m_commit_offset) {
+    if (next_height % m_params.commitment_period == m_commit_offset) {
```
a4191b57b21a0fea871206f368972c7741ad6437 is a bit complicated, can you split it up to smaller refactors?

l0rinc reviewed (changes requested):
Finished the review, my main concern is that the last commit is a bit hard to review (or that I got tired by the end). Left a few more nits as well.

marcofleon reviewed (commented):
Code review ACK a4191b57b21a0fea871206f368972c7741ad6437

I made sure the new unit test doesn't omit any of the current test's coverage. Breaking it into the three functions (sneaky, happy, and low work) makes this fairly easy to confirm. The `HappyPath` case adds coverage for when we get close to the redownload buffer, making sure that no headers are added to the block index until we cross that threshold.

Ran `headerssync-params.py` to verify that the new params added in `chainparams.cpp` are correct. 

Just some small nits below. I also was thinking that it could be good to have the headers sync fuzz test choose from a reasonable range of different commitment periods and redownload buffers. The current one always uses the mainnet params, which is fine too.

hodlinator reviewed (commented):
Thanks for joining @marcofleon! :)

Notable changes in latest push:
* Extracted most unit test-changes from final commit into later commit (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2271396010).
* Broke out `max_seconds_since_start` from `m_max_commitments` calculation (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2271386176).
* Zero-initialized `HeadersSyncParams`-fields (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2273708568)
* Made `HeadersSyncParams` values fuzzed now as I was pushing anyway (https://github.com/bitcoin/bitcoin/pull/32579#pullrequestreview-3116384902).

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -83,8 +83,18 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
 
     const auto genesis{Params().GenesisBlock()};
 
+    // Block header hash target is half of max uint256 (2**256 / 2), expressible
+    // roughly as the coefficient 0x7fffff with the exponent 0x20 (32 bytes).
+    assert(genesis.nBits == 0x207fffff);
+    // ...which implies around every 2nd hash attempt should succeed...
+    assert(GetBlockProof(CBlockIndex(genesis)) == 2);
+    // ...and that is where we get our minimum PoW threshold.
+    assert(CHAIN_WORK == TARGET_BLOCKS * 2);
```
I don't think it's useful, it's literally just stating:
```C++
constexpr arith_uint256 CHAIN_WORK{TARGET_BLOCKS * 2};
...
assert(CHAIN_WORK == TARGET_BLOCKS * 2);
```

hodlinator replied:
Ideally the genesis-block and the other expressions would be `constexpr`, so these could be `static_assert`s right after `CHAIN_WORK` and `TARGET_BLOCKS`. But given they are 90+ lines apart, I'd rather keep it like this. Hm.. I could experiment with ways of decreasing that line distance if I re-push (extracting function bodies and putting them later, moving these constants after the initial functions).

hodlinator replied:
Moved up the macro above the constants in the latest push, and brought the assert block up closer to the constants - in the process removing the final assert.

l0rinc reviewed (approved):
tested ACK 8fc783a95e568c8a5797da1ac6a9922819863175 (edit: rebased locally to check for conflicts with e.g. https://github.com/bitcoin/bitcoin/pull/33116, but all tests are passing. I also ran a `-reindex` until 400k blocks and all seems fine)

Generalized the header sync params for each chain, split the complicated test into smaller, less stateful chunks, updated testing for current mainchain heights, modernized surrounding code - in small, focused, easy to follow, well commented commits.

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -1,146 +1,239 @@
+
+// Subtract MAX_HEADERS_RESULTS (2000 headers/message) + an arbitrary smaller
+// value (123) so our redownload buffer is well below the number of blocks
+// required to reach the CHAIN_WORK threshold, to behave similar to like mainnet.
```
Cheers, fixed in latest push!

marcofleon reviewed (commented):
Re ACK a91a5f0070129192dc50ed50452a2788d69f2d97

l0rinc commented:
reACK a91a5f0070129192dc50ed50452a2788d69f2d97

hodlinator commented on src/test/fuzz/headerssync.cpp:
```diff
@@ -67,7 +68,11 @@ FUZZ_TARGET(headers_sync_state, .init = initialize_headers_sync_state_fuzz)
+            .commitment_period = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(1, Params().HeadersSync().commitment_period * 2),
+            .redownload_buffer_size = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, Params().HeadersSync().redownload_buffer_size * 2),
+        },
+        /*commit_offset=*/fuzzed_data_provider.ConsumeIntegralInRange<size_t>(1, 1024),
```
Thanks! Pushed almost the same.

hodlinator reviewed (commented):
Sorry for invalidating the ACKs. Implemented suggestion in https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2276448044
* Fixes fuzzing of `commitment_offset` to have min value of 0 instead of 1 as on master, and sets the max depending on the fuzzed `commitment_period`.

marcofleon commented:
Re-ACK 342359f9538874a87b86b67dd98f81cb78237a1e

Nice, everything lgtm. Just some basic cleanups and changing the headers sync fuzz target to choose the params since my original review.

hodlinator committed:
refactor(headerssync): Extract test constants ahead of breakup into functions

Made arith_uint256 constexpr-constructible so it can be used for compile time constants.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

hodlinator committed:
refactor(test): Break up headers_sync_state

Helps logically separate the scenarios being tested.

Also adds missing comment for part 4.

(unique_ptrs and ProcessingResults will be cleaned up in next commit).

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -73,6 +79,17 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
 
+    SneakyRedownload(chain_start, first_chain, second_chain);
+    HappyPath(chain_start, first_chain);
+    TooLittleWork(chain_start, second_chain);
```
Pushed version with multiple test cases.

hodlinator reviewed (commented):
Latest push:
* Breaks up test into multiple boost test cases instead of regular functions (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2277678459). In doing so, I changed back to using a custom fixture as on master. Diffing with dimmed-zebra helps show what was just moved and what actually changed `git diff 342359f 53341ea`.
* Removed redundant `assert()` after bringing that block closer to constants (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2271294627).
* Minor fixups to comments and local variables so that they are changed less and in more proper commits.

marcofleon commented:
Re ACK 53341ea10dc2f7df371b416060863bbc094b8773

Breaking up the three scenarios into their own test cases makes sense. I checked to make sure the test logic is unchanged.

davidgumberg commented on src/kernel/chainparams.h:
```diff
@@ -61,6 +61,15 @@ struct ChainTxData {
     double dTxRate;   //!< estimated number of transactions per second after that timestamp
 };
 
+//! Configuration for headers sync memory usage.
+struct HeadersSyncParams {
+    //! Distance in blocks between header commitments.
+    size_t commitment_period{0};
+    //! Minimum number of validated headers to accumulate in the redownload
+    //! buffer before feeding them into the permanent block index.
+    size_t redownload_buffer_size{0};
+};
+
```
It really feels like the default constructor for this should be deleted, but that's not possible because `CChainParams()` default constructs it: https://github.com/bitcoin/bitcoin/blob/53341ea10dc2f7df371b416060863bbc094b8773/src/kernel/chainparams.h#L165

davidgumberg replied:
Ah, I missed that at least one of the params gets checked for not being default-constructed here:

https://github.com/bitcoin/bitcoin/blob/53341ea10dc2f7df371b416060863bbc094b8773/src/headerssync.cpp#L20

davidgumberg commented on src/test/fuzz/headerssync.cpp:
```diff
@@ -65,9 +66,14 @@ FUZZ_TARGET(headers_sync_state, .init = initialize_headers_sync_state_fuzz)
     start_index.phashBlock = &genesis_hash;
 
+    const HeadersSyncParams params{
+        .commitment_period = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(1, Params().HeadersSync().commitment_period * 2),
+        .redownload_buffer_size = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, Params().HeadersSync().redownload_buffer_size * 2),
```
why limit these to `* 2`?

davidgumberg commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -1,146 +1,246 @@
+
+    // Feed in remaining headers, meeting the work threshold again and
+    // completing the REDOWNLOAD phase:
+    CHECK_RESULT(hss.ProcessNextHeaders({first_chain.begin() + REDOWNLOAD_BUFFER_SIZE + 1, first_chain.end()}, true),
```
nit:

```suggestion
    CHECK_RESULT(hss.ProcessNextHeaders({first_chain.begin() + REDOWNLOAD_BUFFER_SIZE + 1, first_chain.end()}, false),
```

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -1,146 +1,246 @@
-    GenerateHeaders(first_chain, target_blocks-1, Params().GenesisBlock().GetHash(),
-            Params().GenesisBlock().nVersion, Params().GenesisBlock().nTime,
-            ArithToUint256(0), Params().GenesisBlock().nBits);
+    // Pretend the message is still "full", so we don't abort.
```
Base version for this PR: https://github.com/bitcoin/bitcoin/blob/fa05a726c225dc65dee79367bb67f099ae4f99e6/src/test/headers_sync_chainwork_tests.cpp#L95-L98

That comment would indeed be more useful before the first `ProcessNextHeaders`-call. Will correct if I push again.

danielabrozzoni reviewed (commented):
Code Review ACK 53341ea10dc2f7df371b416060863bbc094b8773

I reviewed the code and checked that tests were passing locally, but didn't do any extensive testing.

This PR makes the `HEADER_COMMITMENT_PERIOD` and `REDOWNLOAD_BUFFER_SIZE` configurable by the tests, instead of using an hardcoded value, to avoid ending up testing an unrealistic behavior once `REDOWNLOAD_BUFFER_SIZE` surpasses 15_000 in 0.30.

(I like how we can test what could happen using the first commit and investigating the weird behavior!)

Additionally, the PR makes it easier to follow the `headers_sync_chainwork_test`, by cleaning up code and expanding comments. It also makes the test more extensive, by making sure to check every parameter of ProcessingResult using the `CHECK_RESULT` macro.

davidgumberg commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -1,146 +1,246 @@
-    std::vector<CBlockHeader> first_chain;
-    std::vector<CBlockHeader> second_chain;
+    const auto& first_chain{FirstChain()};
+    const auto& second_chain{SecondChain()};
```
I don't think we should be that worried about a 1 in 2^37 chance of spurious test failure, but we could just double the number of headers if that's a concern.

davidgumberg commented:
crACK fa3d88ac35de75af

The test changes here look good to me. I left some nits that can be addressed if following-up or rebasing. I sanity checked that a fresh node doing headers syncing for mainnet, testnet, and signet work on my machine with this branch.

I think it would be good to have input from one of the authors/reviewers of #25717 on the non-test changes to headers sync in this PR, which are https://github.com/bitcoin/bitcoin/pull/32579/commits/4aea5fa3690310bfc2617e3bf31f555dae7d9337 , and https://github.com/bitcoin/bitcoin/pull/32579/commits/1771e59f1a1fcaa6a1d79c93bc0c0c3b1003cf82.

achow101 commented:
I generally disagree with the approach in this PR. It's doing a lot of refactoring that feels too much like refactoring for the sake of refactoring. I think this can be dropped from the milestone.

hodlinator commented:
@achow101:

> I generally disagree with the approach in this PR. It's doing a lot of refactoring that feels too much like refactoring for the sake of refactoring. I think this can be dropped from the milestone.

### Commit order

**I understand that given the current order of commits, the test refactorings seem less motivated.** Order used to be refactorings + main commit + illustrative test + additional log test (https://github.com/bitcoin/bitcoin/compare/37405238a8c4fe4dff82781c91fb668391eb51d4~9...37405238a8c4fe4dff82781c91fb668391eb51d4). Having all the refactoring first got the tests in better shape so that later behavior changes could be expressed clearer.

@fanquake wrote on IRC 2 weeks ago (https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2025-08-07#1142169;):

> would be good if it's clearer why #32579 needs to go into 30, or not

That made me think it would be better to have an independent commit at the beginning of the PR which breaks once the `REDOWNLOAD_BUFFER_SIZE` is increased. If you prefer the order: all refactorings + illustrative test + main commit, that also makes sense.

### Context of headers sync changes

I also must admit that the refactoring began in the context of my other WIP changes for `HeadersSyncState` - adding a cache (https://github.com/l0rinc/bitcoin/pull/3). Before proposing that one on the main repo, I'm also working on another change to make the number of parallel `HeadersSyncState` instances in *net_processing.cpp* more predictable, based off an old comment from 2022 (https://github.com/bitcoin/bitcoin/pull/25720#discussion_r936707629). Maybe I should have a tracking issue to show my larger interest in this area, rather than this PR appearing like a random drive-by.

### Refactoring direction

If you think the test refactorings go in arbitrary directions, I would be curious to hear which parts you still disagree with now that I've used multiple test cases as you suggested (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2277678459).

## Finishing thoughts

If you think that the changes in the main commit 1771e59f1a1fcaa6a1d79c93bc0c0c3b1003cf82 are also refactoring for the sake of refactoring, rather than parameters ending up where they should be to enable better testing, that's a different conversation.

hodlinator committed:
doc(test): Improve comments

+ new assert helping explain why CHAIN_WORK == TARGET_BLOCKS * 2.

hodlinator committed:
test(headerssync): headers_sync_chainwork test improvements

Introduces CHECK_RESULT for consistently validating ProcessingResult.
* Verifies HeadersSyncState::State directly after ProcessNextHeaders().
* Uses BOOST_REQUIRE_EQUAL for HeadersSyncState::State - Nicer failure output and prevents continuing test in nonsensical state.
* Encourages checking Locator and result.pow_validated_headers.

Changes happy_path to test both full & non-full headers messages.

hodlinator committed:
refactor(test): Store HeadersSyncState on the stack

hodlinator committed:
test(headerssync): Test returning of pow_validated_headers behavior

Adding these checks necessitates increasing the length of the generated test chains so that we can properly exceed the REDOWNLOAD_BUFFER_SIZE during the test.

One can check out this commit and locally revert the TARGET_BLOCKS value change to prove the need for tests being able to control the buffer size, as is done by the next commit. Beyond the current REDOWNLOAD_BUFFER_SIZE of 15'009 we need 3 extra - 15'012 TARGET_BLOCKS:
* 1 for the genesis block.
* 1 for the test wanting to check that we start receiving headers for permanent storage *before* the final header (first_chain.back()).
* 1 to exceed REDOWNLOAD_BUFFER_SIZE in HeadersSyncState::PopHeadersReadyForAcceptance().

(The release process includes an occasional increase of the REDOWNLOAD_BUFFER_SIZE value, see release-process.md and history of headerssync.cpp).

hodlinator committed:
headerssync: Make HeadersSyncState more flexible and move constants

Move calculated constants from the top of src/headerssync.cpp into src/kernel/chainparams.cpp.

Instead of being hardcoded to mainnet parameters, HeadersSyncState can now vary depending on chain or test. (This means we can reset TARGET_BLOCKS back to the nice round number of 15'000).

Signet and testnets got new HeadersSyncParams constants through temporarily altering headerssync-params.py with corresponding GENESIS_TIME and MINCHAINWORK_HEADERS (based off defaultAssumeValid block height comments, corresponding to nMinimumChainWork). Regtest doesn't have a default assume valid block height, so the values are copied from Testnet 4. Since the constants only affect memory usage, and have very low impact unless dealing with a largely malicious chain, it's not that critical to keep updating them for non-mainnet chains.

GENESIS_TIMEs (UTC):
Testnet3: 1296688602 = datetime(2011, 2, 2)
Testnet4: 1714777860 = datetime(2024, 5, 3)
Signet: 1598918400 = datetime(2020, 9, 1)

hodlinator reviewed (commented):
Changes since last push:
* Rebased to resolve conflict with #33274.
* Moved checks for buffering-behavior during redownload from the first commit to after the refactoring commits.
* Since #33274 updated the `REDOWNLOAD_BUFFER_SIZE` constant to surpass `TARGET_BLOCKS` (as predicted), we now have to temporarily increase `TARGET_BLOCKS` for these new checks to pass.
* Switched from temporarily introducing `g_latest_result` and having commit at the end to remove it (53341ea10dc2f7df371b416060863bbc094b8773), to introducing a local copy of the `REDOWNLOAD_BUFFER_SIZE` constant. This means that there is less churn on the lines.
* Added comment about probability of spurious test failure in `sneaky_redownload`-case (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2286383684).
* Moved comment `Pretend the message is still "full", so we don't abort.` within `sneaky_redownload`-case (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2287189020)
* Added coverage for non-full headers messages in `happy_path`-case (https://github.com/bitcoin/bitcoin/pull/32579#discussion_r2286437807)
* Changed `max_seconds_since_start`-constant to use initialization-braces instead of assignment (reducing column width by 2 chars).
* Updated `HeadersSyncParams` for the various chains based upon #33274 and re-running *headerssync-params.py*.

A sense of how the resulting end state of the PR has changed (excluding the rebase) can be viewed by something like:
```shell
meld <(git diff 53341ea10dc2f7df371b416060863bbc094b8773~9..53341ea10dc2f7df371b416060863bbc094b8773) <(git diff eb0f41ee1ed9ea54019fcaa4e3ce33481da4459d~8..eb0f41ee1ed9ea54019fcaa4e3ce33481da4459d)
```

l0rinc reviewed (approved):
ACK eb0f41ee1ed9ea54019fcaa4e3ce33481da4459d

I have checked the diffs after rebase, regenerated and verified the mainchain params, looked for leftovers of decommissioned constants and went over the code again quickly to make sure I still agree with everything after the new changes.

hodlinator commented on src/headerssync.cpp:
```diff
@@ -1,32 +1,26 @@
-    m_commit_offset(FastRandomContext().randrange<unsigned>(HEADER_COMMITMENT_PERIOD)),
+        const HeadersSyncParams& params, const CBlockIndex* chain_start,
+        const arith_uint256& minimum_required_work) :
+    m_commit_offset((assert(params.commitment_period > 0), // HeadersSyncParams field must be initialized to non-zero.
```
What makes a 2-statement lambda better than https://en.cppreference.com/w/cpp/language/operator_other.html#Built-in_comma_operator in this case?

Also not sure why it changes the need for the comment.

hodlinator commented on doc/release-process.md:
```diff
@@ -53,7 +53,7 @@ Release Process
     - Check that the other variables still look reasonable.
   - Run the script. It works fine in CPython, but PyPy is much faster (seconds instead of minutes): `pypy3 contrib/devtools/headerssync-params.py`.
-  - Paste the output defining `HEADER_COMMITMENT_PERIOD` and `REDOWNLOAD_BUFFER_SIZE` into the top of [`src/headerssync.cpp`](/src/headerssync.cpp).
+  - Paste the output defining the header `commitment_period` and `redownload_buffer_size` into the mainnet section of [`src/kernel/chainparams.cpp`](/src/kernel/chainparams.cpp).
```
Thanks for catching this! Pushed fix.

sipa reviewed (commented):
utACK 33d550d3044f9075cc866093c453158288f12dec

hodlinator committed:
refactor(headerssync): Process spans of headers

More lightweight than vectors which needed to be copied in tests. Also good to get rid of headers_batch-vector before breaking up test.

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -92,16 +92,12 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
+    (void)hss->ProcessNextHeaders({{first_chain.front()}}, true);
     // Pretend the first header is still "full", so we don't abort.
-    auto result = hss->ProcessNextHeaders(headers_batch, true);
+    auto result = hss->ProcessNextHeaders(std::span{first_chain}.last(first_chain.size() - 1), true);
```
yeah, that confused me as well - if you need to touch again, we could add a [name hint](https://en.cppreference.com/w/cpp/container/span/subspan) to it:
```C++
std::span{first_chain}.subspan(/*offset=*/1)
```

hodlinator commented:
Latest push resolves conflict with #33332.

(Also changed PR-title (category) from "headerssync:" to "p2p:" based off https://github.com/maflcko/DrahtBot/blob/14ccfd94bc7cc584e547e09582373cb87a0e4553/webhook_features/config.yml and the fact that #25717 used that topic. Maybe one could advocate for expanding the number of categories, but I'll punt on it for now. Hat-tip to l0rinc for the DrahtBot reference).

l0rinc commented:
reACK cc5dda1de333cf7aa10e2237ee2c9221f705dbd9

The only difference after the rebase is a removal of a `constexpr` in one of the `base_uint` constructors.

> changed PR-title (category) from "headerssync:" to "p2p

nit: the commits still state `headerssync` (fine by me, just noticed)

marcofleon reviewed (commented):
code review ACK cc5dda1de333cf7aa10e2237ee2c9221f705dbd9

I don't have a strong opinion on the refactors or commit order, although I can see how this PR might not be the most straightforward to review. Overall, the core changes in 8fd1c2893e6768223069d8b2fdec033b026cb2eb and cc5dda1de333cf7aa10e2237ee2c9221f705dbd9 lgtm. Making the headers sync params configurable ensures that the new test can cover the redownload buffer boundary logic.

edit: I should clarify that I think the refactor of the unit test in 7b00643ef5f932116ee303af9984312b27c040f1 is a clear improvement. Just no strong opinion on the structure/order of the commits themselves.

danielabrozzoni reviewed (commented):
reACK cc5dda1de333cf7aa10e2237ee2c9221f705dbd9

I checked the diff since my last view (https://github.com/bitcoin/bitcoin/pull/32579#pullrequestreview-3182000020 + subspan nit + https://github.com/bitcoin/bitcoin/pull/32579#issuecomment-3287698877), and read the whole code once again to make sure I still understand what's going on :)
